### PR TITLE
Store quality option and use it on all videos

### LINKED ISSUE #638: Support original file as part of HLS options
With the next round of transcoding improvements, we will not be transcoding to the original resolution if the original file is within bandwidth and fastplay specs. This will save us a ton on transcoded storage since we won't have to double store the data.

The original stream will only be used if there's not the same resolution option available in the HLS feed at this time. This would prevent us from trying super high bitrate files that we know probably won't stream. In the future we can have 2 options - XRES-original XRES-optimized. 

To support this, we'll need to customize our HLS selector and add logic to support auto switching based on buffering and possibly bandwidth. We would need to add logic to mimic HLS type switching between the original and HLS stream. 

To get the original file, you'd call the generateStreamUrl function. This would be labeled as the resolution from the claim metadata (width). 

The Auto option would default to this original stream also. We may want to run a bandwidth check (we weren't sure how to do this before...but that may just be part of trying to stream the file now...) to determine if we should play one of the transcoded resolutions instead. If there's buffering on original stream, the option would still stay on Auto, but we'd try the HLS feed instead and then let it do its magic. 

At some point we may want to try the original stream again when in Auto mode to see if it can be played successfully (maybe try the bandwidth test in background?) - this would be the scenario of a mobile user who is going in and out of good coverage. 

- [ ] Save / use quality option to wallet 
- [ ] Ensure you can switch back and forth

Additional info on Slack: https://odysee-workspace.slack.com/archives/C02G20Z2AEL/p1639512134383700

### LINKED ISSUE #952: Store quality option and use it on all videos
This also includes livestreams

Allow users to setup the default in settings. If default it not available, use the next lowest as default when playback starts. 

Probably don't need to save the one a user sets in a particular video, but maybe we can improve the UX, hint at the setting when they do (snackbar?)

### LINKED ISSUE #1041: Show which video quality is active
It would be nice to know which video quality is being used when the quality is set on "Auto".

Possible solution:
The most simple solution that bumps into my mind is adding a red dot next to the quality in use (for example, when you hover the cursor on the "auto" label, 720p would have a red dot next to it if 720p is in use).

### PR DESCRIPTION
**Pull request by saltrafael**
## Fixes

Issue Number: Closes #952, Closes #1041

Also closes #638 (remaining checkboxes)

### CODE REVIEW

**Comment by tzarebczan** [comment]
- [ ] missing quality selector on livestreams

File: ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js [hunk 1]
```diff
@@ -238,35 +287,36 @@ class HlsQualitySelectorPlugin {
...
       if (!this.player.isLivestream && this.player.currentSrc() !== this.player.claimSrcVhs.src) {
         setTimeout(() => this.swapSrcTo('vhs'));
-
-        if (height !== 'auto') {
-          // -- Re-select quality --
-          // Until we have "persistent quality" implemented, we need to do this
-          // because the VHS internals default to "auto" when initialized,
-          // causing a GUI mismatch.
-          setTimeout(() => this.setQuality(height), 1000);
-        }
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js:269]: ## Switching from _Original_ to _vhs_ ends up in Auto
1. http://localhost:9090/@eevblog:7/eevblog-1466-stanford-solar-power-at:0
2. Select _Original_.
3. Select _144p_ --> it ends up in _Auto_ instead.  The network log is also showing it is fetching 360p.

File: ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js [hunk 2]
```diff
@@ -238,35 +287,36 @@ class HlsQualitySelectorPlugin {
    * @param {number} height - A number representing HLS playlist.
    */
   setQuality(height) {
+    const { doSetDefaultVideoQuality } = this.config;
     const qualityList = this.player.qualityLevels();
 
     // Set quality on plugin
     this._currentQuality = height;
+    doSetDefaultVideoQuality(height);
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js:295]: This is changing the default every time I make a change from the player.

If we want the behavior of "carry forward the latest selection as the new default", then the selector in Settings Page no longer make sense and can be removed.

I think the default should only be changed from the Settings Page, and any changes from the player will just be a transient setting for that specific video. I think this model is simpler (both in code and behavior).

File: ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js [hunk 3]
```diff
@@ -104,11 +106,34 @@ class HlsQualitySelectorPlugin {
...
+        vhs.playlists.master,
+        vhs.systemBandwidth,
+        parseInt(safeGetComputedStyle(vhs.tech_.el_, 'width'), 10) * pixelRatio,
+        parseInt(safeGetComputedStyle(vhs.tech_.el_, 'height'), 10) * pixelRatio,
+        vhs.limitRenditionByPlayerDimensions
+      );
+
+      const quality = selectedBandwidth.attributes.RESOLUTION.height;
+
+      return __('Auto (%quality%) --[Video quality popup. Long form.]--', { quality: quality + 'p' });
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js:131]: Update `app-string` with all edits and additions if possible.

Some tips:
- If it's an edit, check if the existing entry in `app-string` is used elsewhere:
  - If yes, add the string as new entry.
  - If no, edit/replace the existing entry.

File: ui/component/settingDefaultQuality/view.jsx [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  return (
+    <>
+      <fieldset-section>
+        <FormField
+          name="default_video_quality"
+          type="select"
+          onChange={handleSetQuality}
+          disabled={!enabled}
+          value={defaultQuality}
```
Comment (infinite-persistence) [inline ui/component/settingDefaultQuality/view.jsx:45]: image

File: ui/component/settingDefaultQuality/view.jsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  const [enabled, setEnabled] = React.useState<boolean>(Boolean(defaultQuality));
+
+  const valueRef = React.useRef(VIDEO_QUALITY_OPTIONS[0]);
+
+  function handleEnable() {
+    if (enabled) {
+      setEnabled(false);
+      // From enabled to disabled -> clear the setting
+      doSetDefaultVideoQuality(null);
```
Comment (infinite-persistence) [inline ui/component/settingDefaultQuality/view.jsx:23]: image

The 2 controls is based on 1 variable, that usually yields confusing UI, although minor.

Example:
1. Set to "720p" + "enabled"
2. Set to "disabled" and save.
3. Return to the Settings Page later.  I expect to be able to just toggle to "enable" and get "720", but the drop-down always reverts to "Auto".

Perhaps just put the "null/disable" case as just another drop-down entry (say, "Off", since 'null' is invalid as a drop-down, as shown in the previous warning), instead of a checkbox.  If want the checkbox, would probably need another variable.

File: ui/component/settingDefaultQuality/view.jsx [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          onChange={handleSetQuality}
+          disabled={!enabled}
+          value={defaultQuality}
+        >
+          {VIDEO_QUALITY_OPTIONS.map((quality) => {
+            const qualityStr = typeof quality === 'number' ? quality + 'p' : toCapitalCase(quality);
+
+            return (
+              <option key={'quality' + qualityStr} value={quality}>
+                {qualityStr}
```
Comment (infinite-persistence) [inline ui/component/settingDefaultQuality/view.jsx:52]: I think this will miss localization for "Auto" and "Original"

File: static/app-strings.json [hunk 7]
```diff
@@ -2246,5 +2246,11 @@
   "Attach images by pasting or drag-and-drop.": "Attach images by pasting or drag-and-drop.",
   "There was a network error, but the publish may have been completed. Wait a few minutes, then check your Uploads or Wallet page to confirm.": "There was a network error, but the publish may have been completed. Wait a few minutes, then check your Uploads or Wallet page to confirm.",
   "Sports": "Sports",
-  "--end--": "--end--"
+  "--end--": "--end--",
```
Comment (infinite-persistence) [inline static/app-strings.json:2249]: `--end--` needs to be at the end.  
- There are still logic that depends on it being that way indirectly (can't recall where, only remember that it's not obvious)
- lt also reduces an unnecessary diff of adding and removing the last comma in the json

File: static/app-strings.json [hunk 8]
```diff
@@ -2246,5 +2246,11 @@
...
+  "Default Video Quality": "Default Video Quality",
+  "Set a default quality for video playback. If the default choice is not available, the next lowest will be used when playback starts.": "Set a default quality for video playback. If the default choice is not available, the next lowest will be used when playback starts.",
+  "Auto": "Auto",
+  "Original": "Original",
+  "Enable default quality setting": "Enable default quality setting",
+  "Auto (%quality%) --[Video quality popup. Long form.]--": "Auto (%quality%)"
```
Comment (infinite-persistence) [inline static/app-strings.json:2255]: Put `Auto (%quality%) --[Video quality popup. Long form.]--` next to `Auto --[Video quality. Short form]--` if can.  It helps the translator because they go through it one-by-one, and the surrounding entries help to provide context.

File: ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js [hunk 9]
```diff
@@ -3,6 +3,8 @@ import videojs from 'video.js';
 import { version as VERSION } from './package.json';
 import ConcreteButton from './ConcreteButton';
 import ConcreteMenuItem from './ConcreteMenuItem';
+import * as QUALITY_OPTIONS from 'constants/player';
+import { safeGetComputedStyle, simpleSelector } from '../videojs-http-streaming--override/playlist-selectors';
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/plugins/videojs-hls-quality-selector/plugin.js:7]: Is this line still needed?

File: ui/component/viewers/videoViewer/internal/plugins/videojs-http-streaming--override/playlist-selectors.js [hunk 10]
```diff
@@ -15,7 +15,7 @@ import Playlist from './playlist';
  * @param {HTMLElement} el the htmlelement to work on
  * @param {string} the proprety to get the style for
  */
-const safeGetComputedStyle = function(el, property) {
+export const safeGetComputedStyle = function(el, property) {
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/internal/plugins/videojs-http-streaming--override/playlist-selectors.js:18]: Do we still need to export this?

File: ui/component/viewers/videoViewer/view.jsx [hunk 11]
```diff
@@ -386,15 +386,13 @@ function VideoViewer(props: Props) {
     player.on('loadedmetadata', () => restorePlaybackRate(player));
 
     // Override "auto" to use non-vhs url when the quality matches.
```
Comment (infinite-persistence) [inline ui/component/viewers/videoViewer/view.jsx:388]: `// Override "auto" to use non-vhs url when the quality matches.` -- it's no longer the reason we are override it for.
Either:
- Just remove the comment, or
- To avoid complaints of no documentation, maybe edit it to `// Override the "auto" algorithm to post-process the result`.  The reason for overriding can be changed in the future without having to update this comment again.

**Review by infinite-persistence — COMMENTED** [review]
Thanks for making the changes.
I didn't get a chance to run and test it (away with laptop), so some of the comments below might be wrong.

**Comment by tzarebczan** [comment]
- [ ] abbreviate on mobile - maybe we just list 1080 vs 1080p, otherwise it cuts off a bit. Maybe it's worse on smaller screens...will check on iphone once it charges

**Comment by tzarebczan** [comment]
one last thing : treat "original" for livestreams as the max quality.

**Comment by tzarebczan** [comment]
Ty, will ship in the am!
